[RFC accepted] add wrapper for discriminant_value intrinsic

durka opened the pull request:
add wrapper for discriminant_value intrinsic

Wraps the `discriminant_value` intrinsic under the name `std::mem::discriminant`. In order to avoid prematurely leaking information about the implementation of enums, the return value is an opaque type, generic over the enum type, which implements `Copy`, `Clone`, `PartialEq`, `Eq`, `Hash`, and `Debug` (notably not `PartialOrd`). There is currently no way to get the value out excepting printing the debug representation.

The wrapper is safe and can be stabilized soon as per discussion in #24263.

r? @brson

durka commented:
I just realized the wrapper should be safe. Another commit incoming...

sfackler commented on src/libcore/raw.rs:
```diff
@@ -93,7 +93,32 @@
     pub vtable: *mut (),
 }
+
+/// Returns the value of the discriminant for the enum variant in `v`.
```
It is probably worth adding a warning that discriminants may not compare in the correct order if the enum is `#[repr(u64)]` due to overflow.

durka replied:
Will do.

On Tue, Jul 12, 2016 at 1:43 PM, Steven Fackler <notifications@github.com>
wrote:

> In src/libcore/raw.rs
> <https://github.com/rust-lang/rust/pull/34785#discussion_r70485187>:
>
> >  pub struct TraitObject {
> >      pub data: *mut (),
> >      pub vtable: *mut (),
> >  }
> > +
> > +/// Returns the value of the discriminant for the enum variant in `v`.
>
> It is probably worth adding a warning that discriminants may not compare
> in the correct order if the enum is #[repr(u64)] due to overflow.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/34785/files/609c17e8d0ceac437261df2c6b4231225f1faf08#r70485187>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAC3n4IaNbDuCLcSiYjaKILI3yho7yvLks5qU9IrgaJpZM4JKnnt>
> .
>

durka replied:
Wait, I don't understand. You mean if the enum is _not_ `#[repr(u64)]`?

On Tue, Jul 12, 2016 at 1:46 PM, Alex Burka <durka42@gmail.com> wrote:

> Will do.
>
> On Tue, Jul 12, 2016 at 1:43 PM, Steven Fackler <notifications@github.com>
> wrote:
>
>> In src/libcore/raw.rs
>> <https://github.com/rust-lang/rust/pull/34785#discussion_r70485187>:
>>
>> >  pub struct TraitObject {
>> >      pub data: *mut (),
>> >      pub vtable: *mut (),
>> >  }
>> > +
>> > +/// Returns the value of the discriminant for the enum variant in `v`.
>>
>> It is probably worth adding a warning that discriminants may not compare
>> in the correct order if the enum is #[repr(u64)] due to overflow.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rust-lang/rust/pull/34785/files/609c17e8d0ceac437261df2c6b4231225f1faf08#r70485187>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/AAC3n4IaNbDuCLcSiYjaKILI3yho7yvLks5qU9IrgaJpZM4JKnnt>
>> .
>>
>
>

sfackler replied:
Er, yeah. Your doc change is what I was thinking of.

alexcrichton commented:
I might personally prefer this being located in the `std::mem` module as opposed to `std::raw`. We've *almost* deleted the entire `std::raw` module and I'd personally like to see it go out all the way!

Also, could the actual stabilization be deferred to a later date? The libs team tends to not apply stabilizations mid-cycle but instead places issues into final comment period at the end of which we decide on the stabilization outcome (as a result of the discussion, if any, that happened).

cc @rust-lang/libs, any objections on moving this to FCP though? Seems good to me!

durka commented:
OK, I had previously asked at the tracking issue if this should be insta-stable (since it is basically stabilizing the intrinsic, even though it doesn't actually do that) and @brson said yes. But I'm happy to mark it unstable for now. I'll move it to `std::mem` as well (though it doesn't really fit there).

durka commented:
@sfackler's comment was lost, but I added a note about comparisons.

alexcrichton commented on src/libcore/mem.rs:
```diff
@@ -611,3 +611,35 @@ pub const POST_DROP_USIZE: usize = POST_DROP_U64 as usize;
+/// enum Foo { A(String), B(i32) }
+///
+/// # fn main() {
+/// assert!(mem::discriminant(Foo::A("bar")) != mem::discriminant(Foo::A("baz")));
```
I think these examples may not compile, the discriminants should be the same here, right? Additionally, shouldn't a shared reference be passed in?

durka replied:
Yeah, I just noticed that and fixed it.

On Tue, Jul 12, 2016 at 2:19 PM, Alex Crichton <notifications@github.com>
wrote:

> In src/libcore/mem.rs
> <https://github.com/rust-lang/rust/pull/34785#discussion_r70492383>:
>
> > +/// same order as actual enum variants would under `#[derive(PartialOrd)]`.
> > +///
> > +/// If `T` is not an enum, the return value is unspecified.
> > +///
> > +/// # Example
> > +///
> > +/// This can be used to compare enums that carry data, while disregarding
> > +/// the actual data:
> > +///
> > +/// ```
> > +/// use std::mem;
> > +///
> > +/// enum Foo { A(String), B(i32) }
> > +///
> > +/// # fn main() {
> > +/// assert!(mem::discriminant(Foo::A("bar")) != mem::discriminant(Foo::A("baz")));
>
> I think these examples may not compile, the discriminants should be the
> same here, right? Additionally, shouldn't a shared reference be passed in?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/34785/files/3a88d8811ae23af836e8ed4349e4a96656303878#r70492383>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAC3n1nIqX5ydVTZ2ue3rNp94IPEA-khks5qU9qggaJpZM4JKnnt>
> .
>

alexcrichton commented on src/libcore/mem.rs:
```diff
@@ -611,3 +611,35 @@ pub const POST_DROP_USIZE: usize = POST_DROP_U64 as usize;
+/// the discriminant values returned by this function will not compare in the
+/// same order as actual enum variants would under `#[derive(PartialOrd)]`.
+/// 
+/// If `T` is not an enum, the return value is unspecified.
```
Could this clarify though that it's not like undefined behavior will happen, it's just that the actual value returned cannot be relied on?

alexcrichton replied:
It may also be worth explaining that the returned value only really makes sense to compare with other discriminants generated from values of the same type? That is, comparing the discriminants of types `A` and `B` won't result in anything meaningful

durka replied:
> Could this clarify though that it's not like undefined behavior will happen, it's just that the actual value returned cannot be relied on?

Done.

> It may also be worth explaining that the returned value only really makes sense to compare with other discriminants generated from values of the same type? That is, comparing the discriminants of types A and B won't result in anything meaningful

Not sure about this one. You can set discriminants by hand, so you could for example set up certain error enums to have known discriminants corresponding to the errors produced by a -sys crate or something.

sfackler commented:
@alexcrichton I want to stabilize type_name or something like it, so we might want to keep std::raw around for things like that.

durka commented:
Is it the case that a discriminant can never be _larger_ than `u64` (I think there is an RFC to add `u128` and what not...)?

durka commented:
For the record (on the theme of my previous comment) I still think it would be better to use the type system to make this function's return type better. But I also understand we want to stabilize something.

alexcrichton commented:
@sfackler hm that's a good point about `type_name`, but I figure we can probably figure out a better way to do that than putting it in `std::raw`

aturon commented:
I wasn't initially a fan of `std::mem` for this, but I can sort of see it now. E.g. this is not *so* different from things like `size_of`, which are type-generic functions giving repr details. I could see `type_name` living here eventually, too.

tbu- commented on src/test/run-pass/discriminant_value-wrapper.rs:
```diff
@@ -0,0 +1,71 @@
+    assert_eq!(mem::discriminant(&NullablePointer::Something(&CONST)), 0);
+
+    // unlike the test for the intrinsic, there are no tests with non-enums here
+    // because the result is documented to be unspecified
```
You could still call it on a struct, just to check that it doesn't panic.

durka replied:
Done.

ollie27 commented on src/libcore/mem.rs:
```diff
@@ -611,3 +611,38 @@ pub const POST_DROP_USIZE: usize = POST_DROP_U64 as usize;
+
+/// Returns the value of the discriminant for the enum variant in `v`.
+///
+/// The returned discriminant is cast to `u64`, no matter what the actual
```
I would like to see more explanation about where the value comes from, along with examples if possible.

Also what stability guarantees do the returned values have. For example is `discriminant_value(&Some(123)) == 1` guaranteed to always be `true`?

durka replied:
Added an example.

> Also what stability guarantees do the returned values have.

Hmm... this is a tough one. Since the discriminant depends on the order of enum variants, stability would imply that libraries following semver may not shuffle the order of variants, or add new ones except at the end, in minor versions (actually, they already can't add new variants, unless they have a hack like `std::io::ErrorKind::__Nonexhaustive`). This seems OK to me (needs to be documented of course), but others might have thoughts?

sfackler replied:
I think we should say that consumers shouldn't depend on discriminants being stable across versions of a dependency unless otherwise documented on the enum.

I think it's probably reasonable to state that discriminants will be stable for a given enum definition though across compilations. Maybe compiler versions as well?

durka replied:
Quoth the [reference](https://doc.rust-lang.org/reference.html#enumerations)

> If a discriminant isn't specified, they start at zero, and add one for each variant, in order.

So I think they must be stable across compiler minor versions.

nagisa replied:
>  If a discriminant isn't specified, they start at zero, and add one for each variant, in order.

This is for C-like (i.e. numeric/data-less) enums only. We do not specify anything for sum-types. I would like to not guarantee anything other than `discriminant(Enum::Variant) == discriminant(Enum::Variant)` except for C-like enums, where the usual increment-by-one rules apply.

nagisa replied:
Seems problematic with `#[repr(i/u128)]` potentially happening once we get `i/u128` primitives. Are we gonna forbid such attribute?

durka replied:
I agree. How likely are we to add those primitives?

durka replied:
> This is for C-like (i.e. numeric/data-less) enums only.

The reference does not restrict or qualify this statement in any way. If you want to remove this as a guarantee for all enums, then change the reference. I do understand the reluctance to make guarantees, but this is a bit ridiculous. What would the motivation be for changing discriminants according to the phase of the moon?

ollie27 commented on src/libcore/mem.rs:
```diff
@@ -611,3 +611,38 @@ pub const POST_DROP_USIZE: usize = POST_DROP_U64 as usize;
+///
+/// # Example
+///
+/// This can be used to compare enums that carry data, while disregarding
```
Can it be used for anything else? If so it would be nice to have some examples.

durka replied:
Have any ideas?

durka commented:
Need some more visibility on [this question](https://github.com/rust-lang/rust/pull/34785#discussion_r70613079).

aturon commented:
cc @rust-lang/compiler

durka commented:
I added some [stability docs](https://github.com/rust-lang/rust/pull/34785/files#diff-63525c24a93da7c42c9f2b9a12620342R626).

nagisa commented on src/libcore/mem.rs:
```diff
@@ -611,3 +611,80 @@ pub const POST_DROP_USIZE: usize = POST_DROP_U64 as usize;
+///
+/// If `T` is not an enum, the return value is unspecified (this is **not** undefined
+/// behavior, but the defined behavior is that the value is unpredictable and subject
+/// to change).
```
Feels like not a proper place to explain what a difference between unspecified value and undefined behaviour is. I feel like the most of the text between parenthesis is not necessary and the paragraph could instead read just:

> If `T` is not an enum, the return value is unspecified, but does not result in undefined behaviour.

durka replied:
I added this because it was requested: https://github.com/rust-lang/rust/pull/34785#discussion_r70492519

nagisa replied:
Yeah, the request was to clarify that undefined behaviour will not happen. I feel like what I quoted covers that well enough without trying to explain actual difference between the two in the simple way¹.

¹: A difference between these is a bag of snakes and such explanations are more harmful than useful.

nagisa commented on src/libcore/mem.rs:
```diff
@@ -611,3 +611,80 @@ pub const POST_DROP_USIZE: usize = POST_DROP_U64 as usize;
+/// # Stability
+///
+/// Discriminants can change if enum variants are reordered, if a new variant is added
+/// in the middle, or (in the case of a C-like enum) if explicitly set discriminants are changed.
```
As previously, I would like to not guarantee anything about discriminants in sum-types. i.e. it seems to me that it would be completely fine if discriminants changed even between recompilations of the crate with the same version of the Rust compiler as long as the `discriminant(Enum::Value) == discriminant(Enum::Value)` holds for all dependencies etc.

nagisa commented on src/libcore/mem.rs:
```diff
@@ -611,3 +611,80 @@ pub const POST_DROP_USIZE: usize = POST_DROP_U64 as usize;
+///     X,                      // discriminant: 0
+///     Y(i32),                 // discriminant: 1
+///     Z(Option<&'static str>) // discriminant: 2
+/// }
```
This is leaking-documenting the “guarantees” about discriminants in sum-types again.

nagisa commented on src/libcore/mem.rs:
````diff
@@ -611,3 +611,80 @@ pub const POST_DROP_USIZE: usize = POST_DROP_U64 as usize;
+/// assert!(mem::discriminant(&Foo::A("bar")) == mem::discriminant(&Foo::A("baz")));
+/// assert!(mem::discriminant(&Foo::B(1))     == mem::discriminant(&Foo::B(2)));
+/// assert!(mem::discriminant(&Foo::B(3))     != mem::discriminant(&Foo::C(3)));
+/// ```
````
This example is great. Exactly what I have in mind when saying we should not guarantee too much.

durka commented:
@nagisa thanks for the detailed comments, I will respond individually this afternoon.

ollie27 commented:
Would it perhaps be sufficient to instead stabilise something like the following instead to avoid having to make complicated guarantees about the return value?

```rust
pub fn eq_discriminant<T>(x: &T, y: &T) -> bool {
    unsafe {
        intrinsics::discriminant_value(x) == intrinsics::discriminant_value(y)
    }
}
```

durka commented:
I believe Servo is the main customer hurrying this along, so we'd have to
see if that satisfies their use case.

On Thu, Jul 14, 2016 at 12:15 PM, Oliver Middleton <notifications@github.com
> wrote:

> Would it perhaps be sufficient to instead stabilise something like the
> following instead to avoid having to make complicated guarantees about the
> return value?
>
> pub fn eq_discriminant<T>(x: &T, y: &T) -> bool {
>     unsafe {
>         intrinsics::discriminant_value(x) == intrinsics::discriminant_value(y)
>     }
> }
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/34785#issuecomment-232714395>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAC3nzhT8KS1M7OUJW7gJX2Ipo6NqayGks5qVmC_gaJpZM4JKnnt>
> .
>

sfackler commented:
We'd want something that returns an `Ordering` additionally/instead.

eddyb commented:
`Ordering` would be an improvement because it can handle 64-bit signedness correctly (for enums with explicit discriminants) and any future extensions (like 128-bit integers @nagisa pointed out).

SimonSapin commented:
> I believe Servo is the main customer hurrying this along

As I mentioned in the RFC, we (Servo) have replaced usage of `disciminant_value` with automatically-generated methods that use a giant `match` expression.

It's not ideal so I'd still like to see something stable eventually, but I think there's no need to hurry this if we can do better. (Maybe a `derive`able trait with an associated type for the discriminant integer type?)

durka commented:
@eddyb how would it handle signedness correctly? There's no way to find out what the actual repr type is.

eddyb commented:
@durka By making it an intrinsic (sadly).

durka commented:
@SimonSapin and what is done with the results of the match? Comparisons or directly storing the values?

durka commented:
> (Maybe a deriveable trait with an associated type for the discriminant integer type?)

Could be. But the derive would just generate a giant match statement, so it wouldn't solve any code bloat/compile time issues.

nagisa commented:
@durka not necessarily. We could by all means implement a version of intrinsic that returns a `T` (ensuring that `T` matches size of the actual discriminant) and the derive then could just call that. The types would still be correct, but I think derive wouldn’t yet have information necessary to generate the appropriate associated type.

durka commented:
What would that type signature look like? I think it'd have to be a trait, like I [suggested](https://github.com/rust-lang/rust/issues/24263#issuecomment-225487140), or wait for `fn discriminant_value<T>(v: &T) -> impl Any` :p

nagisa commented:
I’m envisioning something like

```
mod std { mod intrinsics { extern "rust-intrinsic" { fn discriminant<I, O>(v: &I) -> O } } }

#[derive(HasDiscriminant)]
enum E { ... } 

// expands to
// impl ::std::mem::HasDiscriminant for E {
//     type T = <whatever is appropriate, but how>;
//     unsafe fn discriminant(&self) -> Self::T { ::std::intrinsics::discriminant(self) }
// }
```

This has several benefits: 

First of all it would allow to report an error when `#[derive]` in question is not used with `enum`.

In addition, you could also see which enums opt into such unsafe code at enum declaration, thus possibly preventing programmers from changing the enum or forcing them to investigate every use of discriminant (i.e. whether or not something dumb like `e.discriminant() == 0` is done) before changing the enum.

Finally, it feels to me like the generic code using using `HasDiscriminant`-like trait would become more elegant. Not only you could discriminate¹ on types with specific discriminant types, if you ever wanted to do that, but also, it would clearly mark the functions which are using “the trick”, for more throughout inspection.

¹: I mean “add a bound”… pun might have been intended.

durka commented:
I like that! The only drawback (a significant one) is you'd have to
petition the authors of all your favorite libraries to add the derive.

On Thu, Jul 14, 2016 at 3:38 PM, Simonas Kazlauskas <
notifications@github.com> wrote:

> I’m envisioning something like
>
> mod std { mod intrinsics { extern "rust-intrinsic" { fn discriminant<I, O>(v: &I) -> O } } }
>
> #[derive(HasDiscriminant)]
> enum E { ... }
>
> // expands to
> // impl ::std::mem::HasDiscriminant for E {
> //     type T = <whatever is appropriate, but how>;
> //     unsafe fn discriminant(&self) -> Self::T { ::std::intrinsics::discriminant(self) }
> // }
>
> This has several benefits:
>
> First of all it would allow to report an error when #[derive] in question
> is not used with enum.
>
> In addition, you could also see which enums opt into such unsafe code at
> enum declaration, thus possibly preventing programmers from changing the
> enum or forcing them to investigate every use of discriminant (i.e. whether
> or not something dumb like e.discriminant() == 0 is done) before changing
> the enum.
>
> Finally, it feels to me like the generic code using using HasDiscriminant-like
> trait would become more elegant. Not only you could discriminate¹ on types
> with specific discriminant types, if you ever wanted to do that, but also,
> it would clearly mark the functions which are using “the trick”, for more
> throughout inspection.
>
> ¹: I mean “add a bound”… pun might have been intended.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/34785#issuecomment-232769670>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAC3n_jxZOuLUyjXr0mJwtfMl531amHHks5qVpArgaJpZM4JKnnt>
> .
>

durka commented:
I don't know why `HasDiscriminant::discriminant` would be unsafe, though.

On Thu, Jul 14, 2016 at 3:40 PM, Alex Burka <durka42@gmail.com> wrote:

> I like that! The only drawback (a significant one) is you'd have to
> petition the authors of all your favorite libraries to add the derive.
>
> On Thu, Jul 14, 2016 at 3:38 PM, Simonas Kazlauskas <
> notifications@github.com> wrote:
>
>> I’m envisioning something like
>>
>> mod std { mod intrinsics { extern "rust-intrinsic" { fn discriminant<I, O>(v: &I) -> O } } }
>>
>> #[derive(HasDiscriminant)]
>> enum E { ... }
>>
>> // expands to
>> // impl ::std::mem::HasDiscriminant for E {
>> //     type T = <whatever is appropriate, but how>;
>> //     unsafe fn discriminant(&self) -> Self::T { ::std::intrinsics::discriminant(self) }
>> // }
>>
>> This has several benefits:
>>
>> First of all it would allow to report an error when #[derive] in
>> question is not used with enum.
>>
>> In addition, you could also see which enums opt into such unsafe code at
>> enum declaration, thus possibly preventing programmers from changing the
>> enum or forcing them to investigate every use of discriminant (i.e. whether
>> or not something dumb like e.discriminant() == 0 is done) before
>> changing the enum.
>>
>> Finally, it feels to me like the generic code using using HasDiscriminant-like
>> trait would become more elegant. Not only you could discriminate¹ on types
>> with specific discriminant types, if you ever wanted to do that, but also,
>> it would clearly mark the functions which are using “the trick”, for more
>> throughout inspection.
>>
>> ¹: I mean “add a bound”… pun might have been intended.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rust-lang/rust/pull/34785#issuecomment-232769670>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/AAC3n_jxZOuLUyjXr0mJwtfMl531amHHks5qVpArgaJpZM4JKnnt>
>> .
>>
>
>

nagisa commented:
> I don't know why `HasDiscriminant::discriminant` would be unsafe, though.

Ah right, I don’t care either way. Same about the trait. It could be also safe, but I feel like there may be some invariants that should be verified by the implementors if they ever did that.

durka commented:
But yeah, as above, if the repr isn't set explicitly then I don't know how
#[derive] can figure out what the associated type would be.

On Thu, Jul 14, 2016 at 3:43 PM, Simonas Kazlauskas <
notifications@github.com> wrote:

> I don't know why HasDiscriminant::discriminant would be unsafe, though.
>
> Ah right, I don’t care either way. Same about the trait. It could be also
> safe, but I feel like there may be some invariants that should be verified
> by the implementors if they ever did that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/34785#issuecomment-232770924>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAC3n9q2uR6MerbcMqXpDF9Bt8KC6jBdks5qVpFugaJpZM4JKnnt>
> .
>

durka commented:
I guess we could make it a requirement that you have to set the repr in
order to derive the trait.

On Thu, Jul 14, 2016 at 3:52 PM, Alex Burka <durka42@gmail.com> wrote:

> But yeah, as above, if the repr isn't set explicitly then I don't know how
> #[derive] can figure out what the associated type would be.
>
> On Thu, Jul 14, 2016 at 3:43 PM, Simonas Kazlauskas <
> notifications@github.com> wrote:
>
>> I don't know why HasDiscriminant::discriminant would be unsafe, though.
>>
>> Ah right, I don’t care either way. Same about the trait. It could be also
>> safe, but I feel like there may be some invariants that should be verified
>> by the implementors if they ever did that.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rust-lang/rust/pull/34785#issuecomment-232770924>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/AAC3n9q2uR6MerbcMqXpDF9Bt8KC6jBdks5qVpFugaJpZM4JKnnt>
>> .
>>
>
>

nagisa commented:
> I like that! The only drawback (a significant one) is you'd have to petition the authors of all your favorite libraries to add the derive.

On the other hand, adding this derive is a very serious commitment to stability, since even the enum with private fields may have their discriminant read.

durka commented:
Enums don't have private variants, only `#[doc(hidden)]` hacks and stability hacks which only libstd can do.

petrochenkov commented:
>Enums don't have private variants

Yet

SimonSapin commented:
> I don't know how #[derive] can figure out what the associated type would be.

Magic!

Isn’t that the kind of thing you get to do when you’re in compiler internals?

durka commented:
Well, sure, but derivings happen before, say, constant evaluation, which is
necessary to figure out the repr type if not explicitly annotated.

On Thu, Jul 14, 2016 at 4:01 PM, Simon Sapin <notifications@github.com>
wrote:

> I don't know how #[derive] can figure out what the associated type would
> be.
>
> Magic!
>
> Isn’t that the kind of thing you get to do when you’re in compiler
> internals?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/34785#issuecomment-232775462>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAC3n0GVOjGW73ZGYEZ6jxoYXIBCkVF5ks5qVpWxgaJpZM4JKnnt>
> .
>

brson commented:
Still seems to be some open questions about what exactly this is guaranteeing, and whether we even need to stabilize it.

nagisa commented:
> There's a whole lot of design discussion going on here when the original decision was to stabilize the intrinsic as-is, with the intent of doing a better design later. We should decide whether we're really doing that and if not close this PR and go back to the drawing board.

I’m personally fine with stabilising the wrapper in its current form (with adjustments to guarantees, of course). Its not like an existence of stable function to retrieve discriminant would prevent us in any way from adopting a more intricate/correct scheme later on.

---

The only real problem we still have is signed-ness mismatch between `u64` and C-like enums using signed discriminants, but I personally see no issue here as long as we do not tell something like 

> discriminant() returns the same number that goes after `=` in `Variant = 42` or `Variant2 = -42`.

as we currently do with the `Giga`/`Mega`/`Kilo` example and instead tell

> discriminant() will always return a number uniquely describing the variant; the number returned may not match the `-42` in `Variant = -42`.

We could perhaps document the negative number case in some `# Corner cases` section of documentation explaining it, while reassuring the reader that the output will still uniquely identify a variant all of the time.

nagisa commented:
My quick take on what documentation should look like:

```
/// Returns a number uniquely identifying the variant of enum value.
///
/// For C-like enumerations, the number returned by this function shall match the discriminant of
/// the variant cast to `u64`.
///
/// Calling this function with non-enum type will not result in undefined behaviour, but the value 
/// returned is unspecified.
///
/// # Stability
///
/// Discriminants for enum `E` returned by this function may change between different compilations
/// of the crate defining `E`.
///
/// # Examples
///
/// Primary use-case for this function is explicit structural comparison.
///
/// ```
/// #![feature(discriminant_value)]
/// use std::mem;
///
/// enum Foo {
///     A(&'static str),
///     B(i32),
///     C(i32)
/// }
///
/// assert_eq!(mem::discriminant(&Foo::A("bar")),
///            mem::discriminant(&Foo::A("qux")));
/// assert_eq!(mem::discriminant(&Foo::B(1)),
///            mem::discriminant(&Foo::B(2)));
/// assert!(mem::discriminant(&Foo::B(3)) != mem::discriminant(&Foo::C(3)));
/// ```
///
/// The number returned might not match the variant discriminant specified in the definition of the
/// enum, but shall be equal to the value of variant discriminant cast to `u64`:
///
/// ```
/// #![feature(discriminant_value)]
/// use std::mem;
///
/// enum Prefix {
///     None,
///     Kilo = 3,
///     Milli = -3
/// }
///
/// assert_eq!(mem::discriminant(&Prefix::Milli), Prefix::Milli as u64); // not -3!
/// assert_eq!(mem::discriminant(&Prefix::None), Prefix::None as u64);
/// assert_eq!(mem::discriminant(&Prefix::Kilo), Prefix::Kilo as u64);
/// ```
```

brson commented:
Thanks @nagisa.

durka commented:
I still would be interested to know the hypothetical motivation for changing the discriminants of enums in the same code compiled twice by the same compiler. So much for reproducible builds I guess.

For the negative number case, if I'm following, the issue is that in C-like enums, discriminants are set (and yes, the [reference](https://doc.rust-lang.org/reference.html#enumerations) specifies that you _are_ setting the discriminant) as `isize` but this function will return a `u64`, so you would have to cast one of them to compare. That is, `Enum::Variant as u64 == discriminant(&Enum::Variant)` and `discriminant(&Enum::Variant) as $actual_repr_type == Enum::Variant as $actual_repr_type` should hold. So I don't think we need to remove the whole guarantee. If in the future we allow `#[repr(u128)]`, then this will become a larger problem.

While we dither here, nobody else has stepped up in the original thread to demand stabilization, and apparently Servo has replaced their use-case with a macro for the moment. The major sticking point seems to be that the return type of simply `u64` doesn't express what we want it to. Is my idea of an "intrinsic trait" with an associated type too ridiculous to consider? Should I write an RFC?

glaebhoerl commented:
As long as we're dithering, and the desired formulation is still up in the air: if we could opt for one of the many formulations which sidesteps the whole parametricity issue (so `fn discriminant_value<T: Enum>` with `Enum` being automatically implemented by all `enum`s; or explicit `#[derive(DiscriminantValue)] enum Foo` as [suggested by](https://github.com/rust-lang/rust/pull/34785#issuecomment-232769670) @nagisa; or just `fn discriminant_value<T: Reflect>`), I would really prefer that. The situation where the loss of parametricity goes into stable by the back door out of expedience while specialization itself is still in unstable bothers me on some level. It does seem quite unlikely that the lang team would change their opinion around this issue, but on the other hand, it would be *really* awkward if, by some chance, they did, but it were too late because of a small thing like `discriminant_value` having already snuck into stable in the meantime.

alexcrichton commented:
The libs team got a chance to talk about this yesterday, and our conclusion was that this seems fine to merge but with a lack of active users desiring this (it appears Servo has moved to something else) it may be worth making this more extensible into the future. To that end we were thinking that this could return an opaque newtype implementing the `PartialEq`, `Clone`, `Copy`, and `Debug` traits. 

Eventually once we've fixed the implementation we could implement traits like `PartialOrd` or perhaps even provide accessors for the value, but the idea is that the newtype will be maximally flexible for now.

sfackler commented:
One thing I just thought of is that we may want to parameterize the return type: `fn discriminant_value<T>(t: &T) -> Discriminant<T>`.

eddyb commented:
@sfackler Indeed, that will allow it to have type-dependent properties in the future.

SimonSapin commented:
To repeat my earlier message, we have a work-around in Servo for the lack of this functionality.

> It's not ideal so I'd still like to see something stable eventually, but I think there's no need to hurry this if we can do better.

----

@alexcrichton 
> To that end we were thinking that this could return an opaque newtype implementing the PartialEq, Clone, Copy, and Debug traits.
>
> Eventually once we've fixed the implementation we could implement traits like PartialOrd

That seems overly conservative to me. One of the things we do with discriminants in Servo is use them as `HashMap` keys, and I’ve considered switching to `BTreeMap`. If there’s an opaque type, why wait to make it implement `Hash`, `Ord`, etc?

alexcrichton commented:
@sfackler good point!

@SimonSapin the position of being overly conservative seemed ok to us as no one seemed to have a *burning desire* for this immediately, but adding something like `Hash` seems fine to stick this in hash maps or something like that. We specifically were trying to avoid `Ord` and/or `PartialOrd` due to @sfackler's earlier concerns

durka commented:
Given that it's going to be something like `struct Discriminant<T>(u64, PhantomData<&'static T>);` for now, we can derive any traits we want. It makes sense to avoid ordering for now because ideally an enum with a signed repr would have a signed inner type here (in the glorious future where it is `struct Discriminant<T: Enum>(T::Repr);`.

durka commented:
If the newtype is supposed to be opaque, should Debug print out the value, or not? I guess we have plenty of precedent for changing Debug output.

brson commented:
> I still would be interested to know the hypothetical motivation for changing the discriminants of enums in the same code compiled twice by the same compiler. So much for reproducible builds I guess.

Can you expand on this? Reproducible builds are definitely going to become a requirement. How does this patch cause discriminant to change from compilation to compilation?

durka commented:
@brson this patch doesn't, by itself. @nagisa wanted to allow for that possibility which seems overly conservative to me (but I resorted to sarcasm, which was uncalled-for).

nagisa commented:
My point is that we should not tie ourselves by promising things we needn’t to *now*, lest it hindered implementation of whatever else we wanted to. When we are ready to promise reproducible builds, we will by extension promise whatever else necessary to do it, which includes stable discriminants

Note, that this problem goes away completely if we implement some scheme where the end user is not able to observe the actual value of the discriminant (cough, `Discriminant<T>`, cough), because the values become opaque.

nikomatsakis commented:
Just my two cents: with respect to parametricity, I am not keen on defining a special `Enum` trait just for this purpose. I'd want to consider traits like that in a separate RFC (though there may be nice uses for it). I personally think it's nice in any case that we allow the function to take arbitrary types (and not just enum types) -- it seems potentially useful, and it's certainly valid to think of structs, ints and the like as "1-variant enums".  A `Reflect` bound seems ok, and -- I believe? -- like something we could remove later?

durka commented:
I've implemented the `Discriminant<T>` option. I adjusted the docs (still not convinced about how often discriminants may change) and add a `T: Reflect` bound.

@nikomatsakis: I wouldn't introduce a special trait just for this, but I could imagine adding ways to work with algebraic types within the type system in the future. If we get a magic trait in the meantime like `trait HasDiscriminant { type Repr; }`, then we can make it work with this function, without adding a new bound, by implementing the trait for non-enums with `()` as the `Repr`.

nikomatsakis commented:
@durka 

> I wouldn't introduce a special trait just for this, but I could imagine adding ways to work with algebraic types within the type system in the future

Agreed.

arielb1 commented:
So @ubsan has a plan to deprecate `as` casts, and that plan includes having `discriminant_value` as a substitute to the `char -> u32`, `bool -> u32` and `C enum -> u32` casts. Unfortunately, `discriminant_value(&'a')` is 0. Fixing this would make it work.

SimonSapin commented:
@arielb1 Since the `From` and `Into` traits are already in the prelude, I think these conversions are better done with `From<bool> for u8` (and other integer types) and `From<char> for u32` implementations.

durka commented:
@arielb1 luckily `discriminant(&'a')` is unspecified with this PR, leaving
that option open.

On Jul 31, 2016 07:14, "Simon Sapin" <notifications@github.com> wrote:
>
> @arielb1 Since the From and Into traits are already in the prelude, I
think these conversions are better done with From<bool> for u8 (and other
integer types) and From<char> for u32 implementations.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

ubsan commented:
@durka This should be an RFC.

durka commented:
As per above comment, **do not merge** without more discussion of whether this new-ish strategy is the right one, and if so whether a new RFC is needed to propose it.

durka commented:
RFC posted at rust-lang/rfcs#1696.

bors commented:
:umbrella: The latest upstream changes (presumably #35906) made this pull request unmergeable. Please resolve the merge conflicts.

durka commented:
Waiting to rebase until the RFC process concludes.

durka commented:
Rebased and updated.

I removed the `Reflect` bound as per the RFC (nobody opined on it during FCP). I'm assuming that trait's FCP will end in deprecation; if not, we could revisit.

I updated the stability section of the documentation to say (again, nobody opined during FCP):

> Discriminants can change if enum variants are reordered, if a new variant is addedin the middle, or (in the case of a C-like enum) if explicitly set discriminants are changed. Therefore, relying on the discriminants of enums outside of your crate may be a poor decision. However, discriminants of an identical enum should not change between subsequent compilations with the same compiler.

aturon commented:
@nagisa Can you take over this review, given that you've already looked at the PR?

nagisa commented on src/libcore/mem.rs:
```diff
@@ -647,3 +652,79 @@ pub fn drop<T>(_x: T) { }
+///
+/// # Stability
+///
+/// Discriminants can change if enum variants are reordered, if a new variant is added
```
As the return value of this wrapper is a opaque value `Discriminant<T>`, this section reads weird. Namely it changing makes little sense because you aren’t supposed to be able to inspect it directly. Rather the stability and behaviour should be described in terms of `PartialEq` and `Hash`.

durka replied:
That's a good point.

nagisa commented on src/libcore/mem.rs:
```diff
@@ -647,3 +652,79 @@ pub fn drop<T>(_x: T) { }
+#[unstable(feature = "discriminant_value", reason = "recently added, follows RFC", issue = "24263")]
+impl<T> fmt::Debug for Discriminant<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(fmt)
```
This probably wants to use a debug builder and output something along the lines of `Discriminant(<number>)`, rather than outputting a plain number.

durka replied:
OK.

nagisa commented on src/libcore/mem.rs:
```diff
@@ -647,3 +652,79 @@ pub fn drop<T>(_x: T) { }
+pub struct Discriminant<T>(u64, PhantomData<*const T>);
+
+#[unstable(feature = "discriminant_value", reason = "recently added, follows RFC", issue = "24263")]
+impl<T> Copy for Discriminant<T> {}
```
Why not derive all these implementations?

durka replied:
You'd get nonsense bounds on `T`.

durka replied:
I thought I'd included a comment to this effect. I'll add one.

durka commented:
I think it's ready, just waiting for tests.

aturon commented:
@durka OK, just give a ping once it's ready for bors and i'll send it along.

durka commented:
Tests and tidy pass locally.

aturon commented:
Huzzah!

@bors: r=nagisa

durka commented:
@bors? you there?

aturon commented:
@bors: r+

aturon commented:
@bors r=nagisa

aturon commented:
@bors: r=nagisa

durka commented:
Should I try reopening as a new PR?

durka commented:
Moved to #36823.
